Content for speculative loading (#29183)
mdn/content: files/en-us/web/html/element/script/type/speculationrules/index.md

---
title: <script type="speculationrules">
slug: Web/HTML/Element/script/type/speculationrules
page-type: html-attribute-value
browser-compat: html.elements.script.type.speculationrules
---

{{HTMLSidebar}}

The **`speculationrules`** value of the [`type`](/en-US/docs/Web/HTML/Element/script/type) attribute of the [`<script>` element](/en-US/docs/Web/HTML/Element/script) indicates that the body of the element contains speculation rules.

Speculation rules take the form of a JSON structure that determine what resources should be prefetched or prerendered by the browser. This is part of the [Speculation Rules API](/en-US/docs/Web/API/Speculation_Rules_API).

## Syntax

```html
<script type="speculationrules">
  // JSON object defining rules
</script>
```

> **Note:** The `src`, `async`, `nomodule`, `defer`, `crossorigin`, `integrity`, and `referrerpolicy` attributes must not be specified.

### Exceptions

- `TypeError`
  - : The speculation rules definition is not a valid JSON object.

## Description

> A `<script type="speculationrules">` element must contain a valid JSON structure that defines speculation rules. The following examples show separate prefetch and prerender rules:

Reviewer: FWIW in importmap I created a specific schema section for some of this https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/importmap#import_map_json_representation

Might not be necessary here, but useful to have a good single link to "reference" for it.

Author: I've added a subheading to cover this, at `/en-US/docs/Web/HTML/Element/script/type/speculationrules#speculation_rules_json_representation`

> **Note:** Consult the [Speculation Rules API](/en-US/docs/Web/API/Speculation_Rules_API) main page for full details on how to use prefetch and prerender effectively.

Each action field contains an array, which in turn contains one or more objects. Each object contains a single rule defining a set of URLs and related parameters.

Specifically, each object can contain the following properties:

- `"source"`
  - : A string representing the source of the URLs to which the rule applies. Possible values are:
    - `"list"`
      - : Denotes that the URLs will come from a specific list.
- `"urls"`
  - : An array of strings representing the list of URLs to apply the rule to. These can be absolute or relative URLs. Relative URLs will be parsed relative to the document base URL (if inline in a document) or relative to the external resource URL (if externally fetched).
- `"requires"` {{optional_inline}}
  - : An array of strings representing capabilities of the browser parsing the rule, which must be available if the rule is to be applied to the specified URLs. Possible values are:

>     - `"anonymous-client-ip-when-cross-origin"`

Reviewer: This feels very google specific. 

Should someone on another browser just never set this?

Reviewer 2: Each browser will need to figure out its own solution. E.g. Safari would likely use https://support.apple.com/en-us/102602 , Firefox might use something related to https://www.mozilla.org/en-US/products/vpn .

This is indeed pretty Google-specific, and I'm unsure what the best documentation strategy for it is.

Author: This is difficult to make a call on, for sure. To make it seem less Chrome-centric, in my next commit I have taken the approach of including some lines about what mechanisms other future implementations might use for this, based on what @domenic wrote above.

Let me know what you think. If this doesn't seem like a good idea, I will just delete the Safari and Firefox lines and got with what's left. It would be irresponsible to not mention the Chrome caveat at all.

Reviewer: I think the lines are OK. Are you really saying that even other chromiums like Edgewill have to work out a solution? (you say that, but then refer to FF and Apple, so I'm not sure).

So to be clear, if I use this code now I would be feature testing for the API so it won't matter for FF/Apple. So I could set `"anonymous-client-ip-when-cross-origin"`. If at some point FF implements the API and uses its own strategy for this part of the spec my assumption is that either this will "make sense" in which case the setting will be obeyed, or it won't make sense and it will be ignored.

Or to put it another way, how do we expect a browser to behave if it gets a nonsensical setting here? (and should we say).

Reviewer 2: > Are you really saying that even other chromiums like Edgewill have to work out a solution? (you say that, but then refer to FF and Apple, so I'm not sure).

Yes, any browser not created by Google cannot send data through a Google proxy server.

> Or to put it another way, how do we expect a browser to behave if it gets a nonsensical setting here? (and should we say).

What do you mean by "nonsensical"? Like `requires: ["asdf"]`?

Reviewer: > Yes, any browser not created by Google cannot send data through a Google proxy server.

Ah, so this is a real question now. 

> What do you mean by "nonsensical"? Like requires: ["asdf"]?

Yes. Or to be more specific, if I set the `"anonymous-client-ip-when-cross-origin"` right now and run the code on Edge browser, what will happen?

Reviewer 2: Since it's a requirement that can't be met, the prefetch will automatically fail. (Documenting this would be good!!)

Author: OK, I've added in a bullet about other Chromium-based browsers having to figure out their own solutions, and I've added in a note about the prefetch failing if this setting is used but a browser can't meet those requirements.

Reviewer: I've proposed an adjustment https://github.com/mdn/content/pull/29183/files#r1342219064

Author: Right. I've put the note about meeting specified requirements at the top, as that makes sense. I've also moved the bullet about other Chromium-based browsers to just below the Chrome one, and updated it to:

> - Other Chromium-based browsers will have to provide their own solutions. Thorough testing in all target browsers is advised.

In terms of adjusting the BCD and remarking that only Chrome supports this at present...well, we don't know that for sure. To conclusively fix this, we'd have to ask all Chromium browsers whether they've implemented some kind of proxy solution to use for this, and update BCD and the page as appropriate. But that will be a real PITA.

For now, I'd like to suggest that we keep this as-is on the page. In the BCD, I'll add in a separate data point just for `requires`, which currently lists it as supported `true` in Chrome, not sure in other Chromium browsers, `false` in Fx/Saf.

We can then ask other vendors for input, and update the info as we find it.

> **Note:** As speculation rules use a `<script>` element, they need to be explicitly allowed in the [`Content-Security-Policy`](/en-US/docs/Web/HTTP/Headers/Content-Security-Policy) [`script-src`](/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src) directive if the site includes it. This is done by adding the `"inline-speculation-rules"` value along with a hash- or nonce-source.

### Further examples

The earlier examples showed separate speculation rules defined for prefetch and prerender. It is possible to define both in a single set of rules:

```html
<script type="speculationrules">
  {
    "prefetch": [
      {
        "source": "list",
        "urls": ["next.html", "next2.html"],
        "requires": ["anonymous-client-ip-when-cross-origin"],
        "referrer_policy": "no-referrer"
      }
    ],
    "prerender": [
      {
        "source": "list",
        "urls": ["next3.html", "next4.html"]
      }
    ]
  }
</script>
```

It is also allowable to include multiple sets of rules in a single HTML file:

```html
<script type="speculationrules">
  {
    "prefetch": [
      {
        "source": "list",

>         "urls": ["next.html", "next2.html"],

Reviewer: I seem to remember somewhere in all the design docs that you might be able to use patterns to indicate URLs rather than lists like this. Did that happen?

Author: That is in the design docs but not yet implemented, or possibly only in an origin trial at the moment. I decided to not document anything until it was at least in Chrome release.

- `"requires"` {{optional_inline}}
  - : An array of strings representing capabilities of the browser parsing the rule, which must be available if the rule is to be applied to the specified URLs. Possible values are:
    - `"anonymous-client-ip-when-cross-origin"`
      - : `"prefetch"`-only. Specifies that the rule matches only if the user agent can prevent the client IP address from being visible to the origin server if a cross-origin prefetch request is issued. Exactly how this works is dependent on browser implementation specifics. For example, Chrome's implementation hides the IP address using a Google-owned proxy, therefore by default it only works for Google-controlled referrers (since in that case, sending the URLs of the destination to Google is not an additional privacy leak). When used on a non-Google-owned site, rules that include this will only match for users that turn on "Enhanced preloading" in `chrome://settings/preloading`.

> - `"referrer"` {{optional_inline}}

Reviewer 2: ```suggestion
- `"referrer_policy"` {{optional_inline}}
```

Author: Updated

>   - : A string representing a specific referrer policy string to use when requesting the URLs specified in the rule — see [`Referrer-Policy`](/en-US/docs/Web/HTTP/Headers/Referrer-Policy) for possible values. The purpose of this is to allow the referring page to set a stricter policy specifically for the speculative request than the policy the page already has set (either by default, or by using `Referrer-Policy`). A laxer policy set in the speculation rules will not override a stricter policy set on the URL itself.

Reviewer 2: ```suggestion
  - : A string representing a specific referrer policy string to use when requesting the URLs specified in the rule — see [`Referrer-Policy`](/en-US/docs/Web/HTTP/Headers/Referrer-Policy) for possible values. The purpose of this is to allow the referring page to set a stricter policy specifically for the speculative request than the policy the page already has set (either by default, or by using `Referrer-Policy`). A laxer policy set in the speculation rules will not override a stricter policy set on the referring page.
```

Author: Updated

      }
    ]
  }
</script>
```

It is also allowable to include multiple sets of rules in a single HTML file:

```html
<script type="speculationrules">
  {
    "prefetch": [
      {
        "source": "list",
        "urls": ["next.html", "next2.html"],
        "requires": ["anonymous-client-ip-when-cross-origin"],
        "referrer_policy": "no-referrer"
      }
    ]
  }
</script>
<script type="speculationrules">
  {
    "prerender": [
      {
        "source": "list",
        "urls": ["next3.html", "next4.html"]
      }
    ]
  }
</script>
```

And multiple rules in a single result set:

```js
<script type="speculationrules">
{
  "prerender": [
    {
      "source": "list",
      "urls": ["one.html"]
    },
    {
      "source": "list",
      "urls": ["two.html"]
    }
  ]
}
</script>
```

### Dynamic rule insertion

Below is an example that feature detects speculation rules and, if supported, dynamically adds a prerender speculation rule via JavaScript:

```js
if (
  HTMLScriptElement.supports &&
  HTMLScriptElement.supports("speculationrules")
) {
  const specScript = document.createElement("script");
  specScript.type = "speculationrules";

>   specRules = {

Reviewer 2: Same un-declared variable as previously.

Author: Updated!

> **Note:** Consult the [Speculation Rules API](/en-US/docs/Web/API/Speculation_Rules_API) main page for full details on how to use prefetch and prerender effectively.

Each action field contains an array, which in turn contains one or more objects. Each object contains a single rule defining a set of URLs and related parameters.

Specifically, each object can contain the following properties:

- `"source"`
  - : A string representing the source of the URLs to which the rule applies. Possible values are:
    - `"list"`
      - : Denotes that the URLs will come from a specific list.
- `"urls"`
  - : An array of strings representing the list of URLs to apply the rule to. These can be absolute or relative URLs. Relative URLs will be parsed relative to the document base URL (if inline in a document) or relative to the external resource URL (if externally fetched).

> - `"requires"` {{optional_inline}}
>   - : An array of strings representing capabilities of the browser parsing the rule, which must be available if the rule is to be applied to the specified URLs. Possible values are:
>     - `"anonymous-client-ip-when-cross-origin"`
>       - : `"prefetch"`-only. Specifies that the rule matches only if the user agent can prevent the client IP address from being visible to the origin server if a cross-origin prefetch request is issued. Exactly how this works is dependent on browser implementation specifics. For example:
>         - Chrome's implementation hides the IP address using a Google-owned proxy, therefore by default it only works for Google-controlled referrers (since in that case, sending the URLs of the destination to Google is not an additional privacy leak). When used on a non-Google-owned site, rules that include this will only match for users that turn on "Enhanced preloading" in `chrome://settings/preloading`.
>         - A future Safari implementation may possibly use something along the lines of [iCloud Private Relay](https://support.apple.com/en-us/102602).
>         - A future Firefox implementation might use something based on the [Mozilla VPN](https://www.mozilla.org/en-US/products/vpn/) product.
>         - Other Chromium-based browsers will have to provide their own solutions.
>           > **Note:** When attempting to use this setting, prefetches will automatically fail in browsers that cannot meet this requirement, even if they support the [Speculation Rules API](/en-US/docs/Web/API/Speculation_Rules_API).
> - `"referrer_policy"` {{optional_inline}}

Reviewer: The issue with requirements not being met is actually a general case, and probably should be up the top. 
I also suggested new line "> In particular, note that `anonymous-client-ip-when-cross-origin` is currently only supported on Chrome "out of the box"." - I am not so sure about this - it's compatibility information outside of BCD. IMO BCD should have a feature for `anonymous-client-ip-when-cross-origin` that shows this is not supported on the other browsers.

```suggestion
- `"requires"` {{optional_inline}}
  - : An array of strings representing capabilities of the browser parsing the rule, which must be available if the rule is to be applied to the specified URLs.
    
     > **Warning:** Prefetches will automatically fail in browsers that cannot meet a listed requirement.
     > In particular, note that `anonymous-client-ip-when-cross-origin` is currently only supported on Chrome "out of the box".
        
    Possible values are:
    
    - `"anonymous-client-ip-when-cross-origin"`
      - : (`"prefetch"`-only.) Specifies that the rule matches only if the user agent can prevent the client IP address from being visible to the origin server if a cross-origin prefetch request is issued.
        Exactly how this works is dependent on browser implementation specifics. For example:
        - Chrome's implementation hides the IP address using a Google-owned proxy, therefore by default it only works for Google-controlled referrers (since in that case, sending the URLs of the destination to Google is not an additional privacy leak). When used on a non-Google-owned site, rules that include this will only match for users that turn on "Enhanced preloading" in `chrome://settings/preloading`.
        - A future Safari implementation may possibly use something along the lines of [iCloud Private Relay](https://support.apple.com/en-us/102602).
        - A future Firefox implementation might use something based on the [Mozilla VPN](https://www.mozilla.org/en-US/products/vpn/) product.
        - Other Chromium-based browsers will have to provide their own solutions.
- `"referrer_policy"` {{optional_inline}}
```

Author: See earlier comment
